ARIA block cipher and ciphersuites

mpg opened the pull request:
## Description

Add support for ARIA cipher (RFC 5794) and associated TLS ciphersuites (RFC 6209).

This replaces Markku's PR https://github.com/ARMmbed/mbedtls/pull/1190 on which it is based. In addition to Markku's work, this PR:
- is rebased on recent developement
- adds tests in `compat.sh` (some require a preview of OpenSSL 1.1.1 to be installed, see our internal document "mbed TLS development environement" for installation instructions)
- enables some ciphersuites that were left half-declared previously
- does the minor cleanups suggested in my ealier review of https://github.com/ARMmbed/mbedtls/pull/1190
- does some size and performance optimisations, while trying to maintain/improve readability
- adapts to new conventions in the library since this work was started

## Status
**READY**

## Requires Backporting
NO  (new feature)

## Migrations
New API added (ARIA module), closely mirroring the API from the AES module.

## Notes

~~Requires CI changes: Before merging, OpenSSL 1.1.1-preX needs to be installed on the CI machines or `all.sh` will fail!~~ Done!

mpg commented:
@yanesca in addition to your normal review, please also review for design and update the labels accordingly if/when you're happy with the design.

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,318 @@
+
+/* BEGIN_CASE */
+void aria_encrypt_ecb( char *hex_key_string, char *hex_src_string,
+                           char *hex_dst_string, int setkey_result )
```
Should we align the arguments here?


```C
void aria_encrypt_ecb( char *hex_key_string, char *hex_src_string,
                       char *hex_dst_string, int setkey_result )
```

If so, there are many other places in this file that need alignment.

mpg replied:
Will fix that and other places.

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,318 @@
+void aria_encrypt_ecb( char *hex_key_string, char *hex_src_string,
+                           char *hex_dst_string, int setkey_result )
+{
+    unsigned char key_str[1000];
```
Why 1000? Add a comment explaining the rationale behind selecting this number, and make a constant if that helps readability. Someone who later attempts to adapt this test to run on a device that doesn't have 4 KiB of stack may want to know why they might not be able to reduce the sizes used here.

Also, consider using heap or globals for these buffers instead of stack for better portability for on-target testing.

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,318 @@
+    mbedtls_aria_context ctx;
+    int key_len, data_len, i;
+
+    memset( key_str, 0x00, 1000 );
```
Consider using `sizeof` instead of `1000`.

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,318 @@
+    memset( output, 0x00, 1000 );
+    mbedtls_aria_init( &ctx );
+
+    key_len = unhexify( key_str, hex_key_string );
```
How do we know 1000 is big enough?

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,318 @@
+    mbedtls_aria_context ctx;
+    int key_len, data_len, i;
+
+    memset(  key_str, 0x00, 1000 );
```
Consider using a single space (instead of a double) after `memset(` for style.

mpg replied:
Will fix.

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,318 @@
+
+    mbedtls_aria_setkey_dec( &ctx, key_str, key_len * 8 );
+    TEST_ASSERT( mbedtls_aria_crypt_cbc( &ctx, MBEDTLS_ARIA_DECRYPT,
+        data_len, iv_str, src_str, output ) == cbc_result );
```
Consider changing how this line is wrapped for readability.

Other functions have similarly awkwardly wrapped lines.

mpg replied:
I'll try to improve the wrapping, though I'm afraid it'll always be slightly awkward.

andresag01 commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+    (b)[(i)    ] = (unsigned char) ( ( (n)       ) & 0xFF );    \
+    (b)[(i) + 1] = (unsigned char) ( ( (n) >>  8 ) & 0xFF );    \
+    (b)[(i) + 2] = (unsigned char) ( ( (n) >> 16 ) & 0xFF );    \
+    (b)[(i) + 3] = (unsigned char) ( ( (n) >> 24 ) & 0xFF );    \
```
Random thought: I wonder what the compiler actually outputs for something like this. For example, in ARMv7-M there is a REV instruction that does exactly this operation. I am curious to know whether its being used. The same for the rotation macros (there is an ROR instruction) used throughout the library...

mpg replied:
I'm don't think `PUT_INT32_LE()` needs `rev` on little-endian platforms: it's basically just a cast (and cast + rev on big-endian platforms). I just checked and GCC and Clang seem to produce a fairly straightforward sequence of shift, or, and ldrb/strb for this and `GET_INT32_LE()`. I think the reason the compiler can't make these a single ldr/str it has not guarantees on how the bytes buffer is aligned. (Whether it's worth special-casing the case where the buffer is aligned probably depends on usage patterns.)

For rotation (in this file, it's `ARIA_P2`) common compilers use the rotation instruction. Tested with GCC and Clang for x86, ARM-GCC, Clang, armcc5 and armcc6 for ARMv6-M (and ARMv8-A except armcc5).

For other byte rearrangements (`ARIA_P3` aka `rev` and `ARIA_P1` aka `rev16`) see [this commit](https://github.com/ARMmbed/mbedtls/pull/1410/commits/fb0e4f0d1a34bc7eca5b87169925c5c086db6c35) for x86 and [that one](https://github.com/ARMmbed/mbedtls/pull/1410/commits/377b2b624d5a5d894965236f0ae7fe8d09813a44) for arm.

mpg commented on include/mbedtls/config.h:
```diff
@@ -48,10 +48,14 @@
  * Used in:
+ *      library/aria.c
  *      library/timing.c
- *      library/padlock.c
```
That's intentional: the "Used in" list if for modules that will use it if available but will work normally otherwise (in that case, typically because there is a pure-C fallback). The "Required by" list is for things that won't compile without it: every entry from that list should have a corresponding check in `check_config.h`.

Padlock was misplaced so I fixed that while I was editing the list to add Aria.

mpg commented on include/mbedtls/config.h:
```diff
@@ -48,10 +48,14 @@
  *
+ * Required by:
+ *      MBEDTLS_AESNI_C
+ *      MBEDTLS_PADLOCK_C
```
I don't think so, see previous comment about "Used in" vs "Required by".

mpg commented on include/mbedtls/config.h:
```diff
@@ -1797,6 +1803,16 @@
+ * Enable the ARIA block cipher (and TLS ciphersuites that use it, if other
+ * requirements for them are met too).
+ *
+ * Module:  library/aria.c
```
Yes it does (as indicated in the parenthesis). I'll add the full list as for Camellia, and a `Caller:` entry too.

mpg commented on include/mbedtls/ssl_ciphersuites.h:
```diff
@@ -169,6 +169,45 @@ extern "C" {
+#define MBEDTLS_TLS_DHE_RSA_WITH_ARIA_128_GCM_SHA256      0xC052 /**< TLS 1.2 */
+#define MBEDTLS_TLS_DHE_RSA_WITH_ARIA_256_GCM_SHA384      0xC053 /**< TLS 1.2 */
+#define MBEDTLS_TLS_ECDHE_ECDSA_WITH_ARIA_128_GCM_SHA256  0xC05C /**< TLS 1.2 */
+#define MBEDTLS_TLS_ECDHE_ECDSA_WITH_ARIA_256_GCM_SHA384  0xC05D /**< TLS 1.2 */
```
No, I don't think there is a reason to break alignment here (we never break it the `0x01`-`0xC03B` range even though it's already quite diverse, so why break it now). I'll remove one space.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+ */
+
+#ifndef GET_UINT32_LE
+#define GET_UINT32_LE(n,b,i)                    \
```
Indeed. Though I guess this was copy-pasted from `aes.c`, I'll change it to our usual whitespace style (but only in `aria.c`).

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+    for (i = 1; i < ctx->nr; i++ )
+        aria_a( &ctx->rk[i][0], &ctx->rk[i][1], &ctx->rk[i][2], &ctx->rk[i][3] );
+
+    return 0;
```
I'll fix that and other occurrences of non-parenthesized `return` statements.

mpg commented on library/ssl_ciphersuites.c:
```diff
@@ -81,6 +81,14 @@ static const int ciphersuite_preference[] =
+    MBEDTLS_TLS_DHE_RSA_WITH_ARIA_256_GCM_SHA384,
+    MBEDTLS_TLS_ECDHE_ECDSA_WITH_ARIA_256_CBC_SHA384,
+    MBEDTLS_TLS_ECDHE_RSA_WITH_ARIA_256_CBC_SHA384,
+    MBEDTLS_TLS_DHE_RSA_WITH_ARIA_256_CBC_SHA384,
```
Regarding SHA-1, the intro of [RFC 6209](https://tools.ietf.org/html/rfc6209) says:

>  The cipher suites with SHA-1 are not included in this document.  Due
   to recent analytic work on SHA-1 [Wang05], the IETF is gradually
   moving away from SHA-1 and towards stronger hash algorithms.

Regarding SHA-256, it's used, but only with ARIA-128, following the common pattern in TLS ciphersuites of couplin the 128-bit version of a cipher with SHA-256 (or earlier SHA-1) and the 256-bit version with SHA-384.

Note that this PR adds all defined ARIA ciphersuites for which we also support the key exchanges (ie we only excluded the DSS, static DH and DH_anon suites).

There is no ARIA-CCM ciphersuite defined in so far in the [iana registry](https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml) so if you wan to use ARIA-CCM in TLS I guess you'll have to write your own RFC :)

mpg commented on include/mbedtls/aria.h:
```diff
@@ -0,0 +1,237 @@
+ * \return         0 if successful, or MBEDTLS_ERR_ARIA_INVALID_KEY_LENGTH
+ */
+int mbedtls_aria_setkey_enc( mbedtls_aria_context *ctx, const unsigned char *key,
+                         unsigned int keybits );
```
Indeed, will fix.

mpg commented on include/mbedtls/aria.h:
```diff
@@ -0,0 +1,237 @@
+/**
+ * \brief          Initialize ARIA context
+ *
+ * \param ctx      ARIA context to be initialized
```
I'll take the AES documentation and s/AES/ARIA/g for all common functions.

mpg commented on include/mbedtls/aria.h:
```diff
@@ -0,0 +1,237 @@
+/**
+ * \file aria.h
+ *
+ * \brief ARIA block cipher
```
Will do.

mpg commented on include/mbedtls/aria.h:
```diff
@@ -0,0 +1,237 @@
+ *
+ * Note: Due to the nature of CTR you should use the same key schedule for
+ * both encryption and decryption. So a context initialized with
+ * mbedtls_aria_setkey_enc() for both MBEDTLS_ARIA_ENCRYPT and MBEDTLS_ARIA_DECRYPT.
```
This will be fixed by aligning with the AES documentation.

mpg commented on include/mbedtls/aria.h:
```diff
@@ -0,0 +1,237 @@
+ *
+ * Note: Due to the nature of CFB you should use the same key schedule for
+ * both encryption and decryption. So a context initialized with
+ * mbedtls_aria_setkey_enc() for both MBEDTLS_ARIA_ENCRYPT and CAMELLIE_DECRYPT.
```
This will be fixed by aligning with the AES documentation.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+ * This implementation is based on the following standards:
+ * [1] http://210.104.33.10/ARIA/doc/ARIA-specification-e.pdf
+ * [2] https://tools.ietf.org/html/rfc5794
+ */
```
I'll add a reference to the standard while switching to the new style.

mpg commented on include/mbedtls/aria.h:
```diff
@@ -0,0 +1,237 @@
+/**
+ * \brief               ARIA-CTR buffer encryption/decryption
+ *
+ * Warning: You have to keep the maximum use of your counter in mind!
```
Indeed this isn't clear at all. I'll replace that with something that makes sense to me based on the requirements of CTR mode. (I'll do so for other block ciphers as well. A shame we have so much duplication IMO.)

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+ * half of App. B.1 in [1] in terms of 4-byte operators P1, P2, P3 and P4.
+ * The implementation below uses only P1 and P2 as they are sufficient.
+ */
+static inline void aria_a( uint32_t *a, uint32_t *b,
```
Good catch, thanks! Will add that.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+ * If we denote the first by of input by 0, ..., the last byte by f,
+ * then inputs are: a = 0123, b = 4567, c = 89ab, d = cdef.
+ *
+ * Reading [1] 2.4 or [2] 2.4.3 in colums and performing simple
```
Thanks, will fix.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+#endif /* MBEDTLS_CIPHER_MODE_CBC */
+
+#if defined(MBEDTLS_CIPHER_MODE_CFB)
+static const uint8_t aria_test2_cfb_ct[3][48] =         // CFB ciphertxt
```
Not that I know of. Will fix.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+#if defined(__GNUC__) && \
+    defined(__i386__) || defined(__amd64__) || defined( __x86_64__)
+/* I couldn't find an Intel equivalent of ret16, so two instructions */
+#define ARIA_P1(x) ARIA_P2( ARIA_P3( x ) )
```
I don't have easy access to MSVC for testing. Ideally we'd like assembly used on most common toolchains and platforms as it brings a 10 to 20% performance boost, but as you say there is a C fallback so I wouldn't say it's a blocker either. Perhaps we should check with the people interested in ARIA what toolchain and platforms they care about, too.

(While at it, having assembly with IAR would be cool too.)

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1022 @@
+ * c = 0022 + 1133 + 4545 + 7667 + ab89 + dcdc + fefe
+ *   = 0123 + 1032 + 4567 + 7654 + ab89 + dcfe + fedc
+ * d = 1001 + 2332 + 6644 + 7755 + 9898 + baba + cedf
+ *   = 1032 + 2301 + 6745 + 7654 + 98ba + ba98 + cdef
```
We don't have a unit test for this function (it's `static` and no test data is available) but we do have tests for setkey+encryption which crucially rely on it, and it seems highly unlikely (like cryptographically unlikely) that we would pass those tests if that function were wrong. Also, due to the nature of the function (just moving bytes around and xoring them, no arithmetic), it seems unlikely for it to be wrong only on some input values.

In addition, the implementation has been symbolically verified (see Python script in the commit message).

Patater replied:
OK. Not a blocker, but how can we be sure the comment is correct and contains no typos, and continues to be correct into the future? I understand that the output shouldn't change over time, and we have no reason to change the comment, but is there anything in place to prevent accidentally changing the comment other than code review? Just curious.

mpg replied:
No, I don't think there is anything other than code review preventing us from making sure this comment is and will remain correct. Each first line is mostly just copied (with obvious substitutions such as `15` -> `f`) from the standard and each second line is a supposedly trivial rearrangement from the first line. I don't think there is anything other than your reviewing work to substantiate the fact that the previous sentence is indeed correct (and future reviews to ensure it remains so).

OTOH, since this is mostly just copied from the standard, we're quite unlikely to want to modify it in the future.

mpg replied:
Aw, that's embarrassing, the comment contained not one but 3 typos. Fix in next commit. Quite surprisingly, it would appear python scripts are better than me at transforming data in a reliable way.

mpg commented:
@Patater I pushed a few commits addressing some of your feedback, replied to a few other comments, and will address the remaining ones later.

Regarding testing, currently the only platforms/toolchains tested by me are Linux/x86_64 (gcc/clang) and Mbed OS (boards base on M4, M3 and M0 cores, using the online compiler). Yesterday I ran `all.sh` on my (Linux) machine and got a few failures related to the assembly that I'll address soon too.

No windows testing was done so far (I just triggered a run on buildbot where we have a windows slave) and not big-endian testing either - I don't have a big-endian (virtual) machine at hand, and neither does the CI so far.

mpg commented:
@patater: I think I've now addressed all comments from your first review. I've also added error codes for hardware acceleration in order to match what's now in AES and all other crypto modules with a `_ALT` option. Finally, I've fixed the two errors found by running `all.sh` on my machine. Please review again.

Patater commented:
@mpg Thanks! I'll have another look.

Patater commented on include/mbedtls/aes.h:
```diff
@@ -300,7 +300,24 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
+ *             2. Use a randomly-generated \p nonce_counter for each call.
+ *             With this strategy, you need to ensure the nonce is generated
+ *             in an unbiased way and you must not encrypt more than 2**64
+ *             block of data.
```
s/block of data/blocks of data

Due to the duplication in our library, this fix must also get duplicated to other block cipher code. :/

mpg replied:
Aw. Fixed.

mpg commented:
Note: `all.sh -kr` passed on my machine.

mpg commented:
Note: some level of windows testing passed on the CI (building with MSVC ind 32 and 64 bit mode, building and running `make test` with mingw32).

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,334 @@
+        for( i = 0; i < data_len; i += MBEDTLS_ARIA_BLOCKSIZE )
+        {
+            TEST_ASSERT( mbedtls_aria_crypt_ecb( &ctx, MBEDTLS_ARIA_DECRYPT,
+            src_str + i, output + i ) == 0 );
```
We probably want this line indented at least a little bit.

mpg replied:
I'm not sure how I missed that one. Thanks.

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,334 @@
+    mbedtls_aria_setkey_enc( &ctx, key_str, key_len * 8 );
+    TEST_ASSERT( mbedtls_aria_crypt_cbc( &ctx, MBEDTLS_ARIA_ENCRYPT, data_len,
+                                         iv_str, src_str, output )
+                 == cbc_result );
```
This is better, thanks.

Patater commented on tests/suites/test_suite_aria.function:
```diff
@@ -0,0 +1,334 @@
+    mbedtls_aria_setkey_enc( &ctx, key_str, key_len * 8 );
+    TEST_ASSERT( mbedtls_aria_crypt_cfb128( &ctx, MBEDTLS_ARIA_ENCRYPT,
+                                            data_len, &iv_offset, iv_str,
+                                            src_str, output ) == result );
```
Minor: Consider having `== result` on a new line for consistency in wrapping style, here an in other places in the file. I'm fine with either style you'd like to go with, but consistency would improve readability.

Previous style example:
```C
TEST_ASSERT( mbedtls_aria_crypt_cbc( &ctx, MBEDTLS_ARIA_DECRYPT, data_len,
                                     iv_str, src_str, output )
             == cbc_result );
```

mpg replied:
Ok.

mpg commented:
@Patater : pushed some new changes, hopefully addressing all of your feedback this time :)

Patater commented on tests/compat.sh:
```diff
@@ -422,6 +430,12 @@ add_common_ciphersuites()
+# Each ciphersuite should appear two times, once with its OpenSSL name, once
+# with its Mbed TLS name.
+#
+# NOTE: for some reason RSA-PSK doesn't work with OpenSSL,
```
This is still an open question. I searched our issues on GitHub for "RSA-PSK", but found nothing. Could you please file an issue if this is indeed still actually an issue?

Patater commented:
@mpg How much Windows testing is left to do?

Please consider filing an issue on GitHub about the RSA-PSS OpenSSL issue.

Thanks

mpg commented:
@Patater Thanks for the reminder about RSA-PSK. I opened this ticket to track it: https://github.com/ARMmbed/mbedtls/issues/1419

Regarding windows testing, I don't have any access to a windows machine (requested a VM from IT some time ago, still pending). If you (or another reviewer or team member) do, I think it would be great if you could build with MSVC (ideally both 32 and 64-bit builds) and then run `test_suite_aria.exe`.

I don't really expect any surprise here, as ARIA code is quite system-independent and the CI already showed it compiled without warnings with MSVC, but of course we never know until we actually run the test.

Patater commented:
I ran Windows tests via our release testing job. They passed.

https://jenkins-internal.mbed.com/view/mbed-tls/job/mbedtls-release/193/execution/node/66/log/

Patater commented on tests/compat.sh:
```diff
@@ -422,6 +430,12 @@ add_common_ciphersuites()
+# with its Mbed TLS name.
+#
+# NOTE: for some reason RSA-PSK doesn't work with OpenSSL,
+# so RSA-PSK ciphersuites need to go in other sections.
```
If you want, you can mention issue #1419 here, to tie everything together. It'll help maintainability.

mpg commented:
Note: just tried running `compat.sh` with OpenSSL 1.1.1-pre2, 184 tests passed, 0 skipped. This is an expected improvement over the results with OpenSSL 1.1.1-pre1 where 8 test were skipped due to an OpenSSL issue that has been fixed in pre2.

sbutcher-arm commented:
@yanesca - Can you please review, and if you approve, remove the 'needs design approval' label, and add the 'approved for design' one? Thanks

JanneKiiskila commented:
@mpg  @Patater  @sbutcher-arm - any word on when this could be merged/released?

yanesca commented on library/aria.c:
```diff
@@ -0,0 +1,1033 @@
+                               const unsigned char *input,
+                               unsigned char *output )
+{
+    int c;
```
I know that this is copied, but I was always puzzled by this so please forgive me for asking: Why don't we declare this as `unsigned char`?

mpg replied:
I have no idea. I think the real reason is that in the past `int` was arguably very overused in the library, and in many places it was never changed, and even a lot of times copied and imitated. I'm moving to `unsigned char` for aria, but not changing it in the other modules so far, as I don't want to over-extend the scope of that PR.

mpg replied:
Btw, I don't think you need to be forgiven for asking, rather perhaps I should ask to be forgiven for _not_ asking when Paul was still around ;)

mpg commented:
@yanesca Thanks for your review! I'll address your comments shortly. Also, if you're satisfied with the design, can you record that by adjusting the labels?

mpg commented on include/mbedtls/aria.h:
```diff
@@ -0,0 +1,315 @@
+ *        parameter (encrypt or decrypt), on the input data buffer
+ *        defined in the \p input parameter.
+ *
+ *        For CFB, you must set up the context with mbedtls_aes_setkey_enc(),
```
How embarrassing. Thanks for noticing. I'll do a search-and-replace.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1033 @@
+ * ARIA Affine Transform
+ * (a, b, c, d) = state in/out
+ *
+ * If we denote the first by of input by 0, ..., the last byte by f,
```
Thanks, fixed.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1033 @@
+ */
+static inline void aria_sl( uint32_t *a, uint32_t *b,
+                            uint32_t *c, uint32_t *d,
+                            const uint8_t sa[0x100], const uint8_t sb[0x100],
```
I tend to agree. I previously left it as Markku wrote it as it's indeed subjective, but now that we're two with the same subjectivity, I changed it.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1033 @@
+        GET_UINT32_LE( w[1][3], key, 28 );
+    }
+
+    i = (keybits - 128) >> 6;               // index: 0, 1, 2
```
Indeed. Fixing, as well as a few other instances.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1033 @@
+    }
+    aria_rot128( ctx->rk[16], w[0], w[1], 19 );
+
+    return( 0 );
```
Ouch, thank you so much for catching that. Fixing.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1033 @@
+ * Encrypt a block
+ */
+int mbedtls_aria_crypt_ecb( mbedtls_aria_context *ctx,
+                            int mode,
```
Indeed we don't. I think it was left here because the API was probably copied from AES. Now that you mention it, I don't think consistency with AES is a good enough reason to have a useless parameter, especially as users will normally never call this function directly. I'm removing the parameter.

mpg commented on library/aria.c:
```diff
@@ -0,0 +1,1033 @@
+#endif /* x86 gnuc */
+#endif /* MBEDTLS_HAVE_ASM && GNUC */
+#if !defined(ARIA_P1)
+#define ARIA_P1(x) ((((x) >> 8) & 0x00FF00FF) ^ (((x) & 0x00FF00FF) << 8))
```
With `-Os` the compiler is likely to decline inlining a `static inline` function here as it would be several instructions long (as opposed to the one-instruction functions using inline assembly, which are always inlined) and is used many times in `aria_a()`. This would have a pretty significant impact on performance as we're in the inner loop of ARIA, which would be IMO too high a price to pay for saving a few bytes of ROM.

I don't remember if I tested exactly that one, but I played quite a bit with a few variants of macros and static functions in a few places, and my conclusion was that sometimes you want to give more than a hint to the compiler that this one _really_ matters for performance and we want it inlined no matter what. I think a macro accomplishes that.

yanesca replied:
Sounds reasonable, thanks for explaining!

mpg commented on include/mbedtls/aes.h:
```diff
@@ -300,7 +300,24 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
+ *
+ *             1. Use a counter starting at 0 or a random value. With this
+ *             strategy, this function will increment the counter for you, so
+ *             you only need to preserve the \p nonce_counter buffer between
```
Indeed. We probably want to mention explicitly that `stream_block` needs to be preserved across calls too then.

mpg replied:
Well I thought about it again and now I think it's a bit more complex, either you want to treat everything as a single stream and then sure you need to preserve `nc_off` and `stream_block`, or conceptually you have a series of messages that could be transmitted individually, and in that second case you're more likely to want to use the second construct mentioned by NIST (see other comment) and then set `nc_off` to `0` for each new message.

I'm afraid explaining only the first case might not be enough for all users. (But then again, should users really use this API directly? And if not, why to we have it?)

yanesca replied:
> But then again, should users really use this API directly?

I have the feeling that this API is too low level for the vast majority of users to use directly, and in the long term I really would like to see it to be changed to something more robust (NIST compliant perhaps?). On the other hand I think that doing it now would be slightly out of scope for this PR ;)

mpg replied:
I agree, and I was not suggesting doing that now :) I'll make a proposal for a new and hopefully better documentation tomorrow.

yanesca replied:
> I'm afraid explaining only the first case might not be enough for all users.

I think that the API we have is flexible enough to implement either versions recommended by NIST:

1. Save `nonce_counter`, `stream_block` and `nc_off` between calls throughout the lifetime of the key and make sure that you don't encrypt more than 2**128 blocks. 
2. Save `nonce_counter`, `stream_block` and `nc_off` between calls while processing a message and make sure that the message is not longer than 2\*\*64 blocks. When starting a new message, increment the first half of `nonce_counter` and reset `nc_off` to zero. Make sure you don't use the key for more than 2**64 messages.

mpg commented on include/mbedtls/aes.h:
```diff
@@ -300,7 +300,24 @@ int mbedtls_aes_crypt_cfb8( mbedtls_aes_context *ctx,
+ *             you only need to preserve the \p nonce_counter buffer between
+ *             calls. With this strategy, you must not encrypt more than
+ *             2**128 blocks of data.
+ *             2. Use a randomly-generated \p nonce_counter for each call.
```
Right, I should have consulted you for review when writing that paragraph. Limiting to `2**64` blocks is absolutely not enough as the probability of collisions becomes non-trivial way before that, suggesting that it's OK to use up to `2**64` random "nonce" values was a stupid mistake of mine.

I did remember the choice between deterministic and random-based construction from the [NIST GCM spec](https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf) (section 8.2) but misremember section 8.3 where NIST obviously doesn't make the same mistake as me and puts the limit at `2**32` invocations with the same key.

For reference, I checked the relevant [NIST document for CTR](https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38a.pdf) and Appendix B.2 describes two strategies, but for the second strategy it does not describe how the nonce should be chosen.

I'm not yet convinced what strategies we should and should not explain, I'll probably keep thinking a bit about it and try to come up with a proposal later.

mpg commented on include/mbedtls/error.h:
```diff
@@ -63,6 +63,7 @@
  * CTR_DBRG  4  0x0034-0x003A
  * ENTROPY   3  0x003C-0x0040   0x003D-0x003F
  * NET      11  0x0042-0x0052   0x0043-0x0045
+ * ARIA      4  0x0058-0x005E
```
I don't know, but since ASN1 uses error codes starting from `0x0060`, the choice is between leaving a small hole before or after ARIA (in place of the larger hole that was present between NET and ASN1 for no apparent reason). I don't think either way is better or worse than the other, so I'd rather leave it unchanged unless you disagree.

mpg commented on library/cipher_wrap.c:
```diff
@@ -822,6 +826,363 @@ static const mbedtls_cipher_info_t camellia_256_ccm_info = {
+    MBEDTLS_MODE_ECB,
+    128,
+    "ARIA-128-ECB",
+    16,
```
We don't need an IV, and hopefully this value is ignored by the code, in which case I agree it would be clearer to set it to 0 or some "signaling" value. But I'm afraid at some point at least some parts of the code tended to use IV size when what they really need to know is the block size. This has probably been fixed the meantime, but unless one of us takes the time to make sure of that, I'd rather leave `16` here at I don't think it's causing harm beyond looking a bit strange.

mpg commented:
@yanesca I think I've addressed all of your feedback (except the two points on the CTR documentation which I'll rework tomorrow) either by fixing the code or explaining why I don't think it's necessary. Feel free to check and let me know i I overlooked something.

I've also added new commits to adjust to new standards set in other PRs, and merged development into this branch in order to resolve conflicts.

mpg commented:
@yanesca I've made an attempt at improving the CTR documentation. If you're happy with it, I'll copy and adapt it to the other modules.

@Patater There has been a few new commits since you approved this PR, could you review them too? Thanks!

mpg commented on include/mbedtls/aria.h:
```diff
@@ -255,19 +255,44 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
  *
  *             There are two common strategies for managing nonces with CTR:
```
Ok. I'll append "with the same key" to each "you should not encrypt more than ..." sentence.

Patater commented:
Could you please fix the Travis CI issues? It looks like there are build failures and Doxygen failures.

mpg commented:
@patater, I think it was all a single error, introduce by a bad edit in a recent commit, that I failed to teset properly as apparently I forgot the feature I'm working on is disabled by default... I just pushed a fix, we'll see how it goes.

Patater commented on include/mbedtls/aria.h:
```diff
@@ -255,19 +255,44 @@ int mbedtls_aria_crypt_cfb128( mbedtls_aria_context *ctx,
  *             Note that for both stategies, the limit is in number of blocks
  *             and that an ARIA block is 16 bytes.
  *
+ * \warning    Upon return, \p stream_block constains sensitive data. Its
```
s/constains/contains

mpg commented:
@Patater Thanks for catching the typo, it made me realise I forgot to copy the warning to other `_ctr` functions that needed it too. Now fixed and propagated.
